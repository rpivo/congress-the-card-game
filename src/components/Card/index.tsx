import React, { memo } from 'react';
import Style from './style';

type CardProps = {
  ability1Title?: string;
  active: boolean;
  id: string;
  handleCardMouseDown: (id: string) => void;
  subtitle: string;
  title: string;
};

const Card = ({
  ability1Title,
  active,
  id,
  handleCardMouseDown,
  subtitle,
  title,
}: Readonly<CardProps>): JSX.Element =>
  <Style
    className={`card${active ? ' active' : ''}`}
    onClick={(event: React.MouseEvent<HTMLInputElement>): void => event.stopPropagation()}
    onMouseDown={(): void => handleCardMouseDown(id)}>
    <p className='title'>{title}</p>
    <p className={'subtext subtitle'}>{subtitle}</p>
    {ability1Title && <p className='subtext ability1Title'>{ability1Title}</p>}
  </Style>;

export default memo(Card);
