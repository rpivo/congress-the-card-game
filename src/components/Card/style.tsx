import styled from 'styled-components';

const Style = styled.div`
  background: #EAEAEA;
  border-radius: 5px;
  cursor: pointer;
  display: inline-block;
  height: 182px;
  padding: 8px;
  place-self: center;
  transition: box-shadow 0.2s ease-in-out;
  width: 130px;

  &.active {
    background: #9FFFB0;
  }

  &.card {
    box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);

    :active {
      background: #9FFFB0;
    }
  
    :hover {
      box-shadow: 0 14px 28px rgba(0,0,0,0.25), 0 10px 10px rgba(0,0,0,0.22);
    }
  }

  &.deck {
    box-shadow: 0 1px 3px rgba(0,0,0,0.72), 0 1px 2px rgba(0,0,0,0.84);
    grid-row: 3; 
    grid-column: 1;
  }
`;

export default Style;
