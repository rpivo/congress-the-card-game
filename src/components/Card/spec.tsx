import { shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import Card from '@components/Card';

describe('App', () => {
  it('should render correctly', () => {
    const tree = renderer
      .create(<Card />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('style', () => {
    it('should default to a background of #EAEAEA', () => {
      const tree = renderer
        .create(<Card />)
        .toJSON();
      expect(tree).toHaveStyleRule('background', '#EAEAEA');
    });

    it('should have a default box-shadow', () => {
      const tree = renderer
        .create(<Card />)
        .toJSON();
      expect(tree).toHaveStyleRule(
        'box-shadow',
        '0 1px 3px rgba(0,0,0,0.12),0 1px 2px rgba(0,0,0,0.24)',
      );
    });

    it('should update background on active', () => {
      const tree = renderer
        .create(<Card />)
        .toJSON();
      expect(tree).toHaveStyleRule(
        'background',
        '#9FFFB0',
        { modifier: ':active' },
      );
    });

    it('should update box-shadow on hover', () => {
      const tree = renderer
        .create(<Card />)
        .toJSON();
      expect(tree).toHaveStyleRule(
        'box-shadow',
        '0 14px 28px rgba(0,0,0,0.25),0 10px 10px rgba(0,0,0,0.22)',
        { modifier: ':hover' },
      );
    });
  });

  describe('classList', () => {
    it('should not initialize with the .active class', () => {
      const wrapper = shallow(<Card />);
      expect(wrapper.find('.card').hasClass('active')).toBe(false);
    });

    it('should update to have the .active class on click', () => {
      const wrapper = shallow(<Card />);
      wrapper.simulate('click');
      expect(wrapper.find('.card').hasClass('active')).toBe(true);
    });
  });
});
