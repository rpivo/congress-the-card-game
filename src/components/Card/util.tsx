import React from 'react';
import Card from '@components/Card';
import cardData from '@utilities/cards.json';

type CardType = {
  [id: string]: {
    abilities?: {
      ability1?: {
        title: string;
        description: string;
      }
    };
    name: string;
    group: string;
    type: string;
  }
};

type GetCardsProps = {
  activeCard?: number,
  cardIDs: number[],
  handleMouseDown: (id: number) => void,
  isHandCard?: boolean,
};

export const getCards = ({
  activeCard,
  cardIDs,
  handleMouseDown,
  isHandCard = false,
}: Readonly<GetCardsProps>): JSX.Element[] => {
  const cards = [];
  for (const id of cardIDs) {
    const { abilities, group, name } = (cardData as CardType)[`${id}`];
    cards.push(
      <Card
        ability1Title={abilities?.ability1?.title}
        ability1Description={abilities?.ability1?.description}
        active={activeCard === id}
        handleCardMouseDown={handleMouseDown}
        id={id}
        isHandCard={isHandCard}
        key={id}
        title={name}
        subtitle={group}
      />
    );
  }
  return cards;
};
