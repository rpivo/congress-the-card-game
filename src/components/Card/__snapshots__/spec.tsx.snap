// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
.c0 {
  background: #EAEAEA;
  border-radius: 5px;
  box-shadow: 0 1px 3px rgba(0,0,0,0.12),0 1px 2px rgba(0,0,0,0.24);
  cursor: pointer;
  display: inline-block;
  height: 182px;
  padding: 8px;
  place-self: center;
  -webkit-transition: box-shadow 0.2s ease-in-out;
  transition: box-shadow 0.2s ease-in-out;
  width: 130px;
}

.c0:active {
  background: #9FFFB0;
}

.c0:hover {
  box-shadow: 0 14px 28px rgba(0,0,0,0.25),0 10px 10px rgba(0,0,0,0.22);
}

<div
  className="c0 card"
>
  Hello
</div>
`;
